pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,322 @@
+	var errs []error
+
+	// Bail out early if updated service is not selectable
+	if !(p.isSelectableService(newSVC) || p.isSelectableService(oldSVC)) {
+		return nil
+	}
+
+	// newService is true if this is the first time we observe this service
+	newService := oldSVC == nil
+	// changedService is true if the service label or selector has changed
+	changedService := !newSVC.DeepEqual(oldSVC)
+
+	// candidatePolicyKeys contains the set of policy names we need to process
+	// for this service update. By default, we consider all policies with
+	// a ToServices selector as candidates.
+	candidatePolicyKeys := p.toServicesPolicies
+	if !(newService || changedService) {
+		// If the service definition itself has not changed, and it's not the
+		// first time we process this service, we only need to check the
+		// policies which are known to select the old version of the service
+		candidatePolicyKeys = p.cnpByServiceID[svcID]
+	}
+
+	// Iterate over all policies potentially affected by this service update,
+	// and re-resolve the policy refs for each.
+	for key := range candidatePolicyKeys {
+		cnp, ok := p.cnpCache[key]
+		if !ok {
+			p.log.WithFields(logrus.Fields{
+				logfields.Key:       key,
+				logfields.ServiceID: svcID,
+			}).Error("BUG: Candidate policy for service update not found. Please report this bug to Cilium developers.")
+			continue
+		}
+
+		// Skip policies which are not affected by this service update
+		if !(p.cnpMatchesService(cnp, svcID, newSVC) ||
+			(!newService && changedService && p.cnpMatchesService(cnp, svcID, oldSVC))) {
+			continue
+		}
+
+		if p.config.Debug {
+			p.log.WithFields(logrus.Fields{
+				logfields.CiliumNetworkPolicyName: cnp.Name,
+				logfields.K8sAPIVersion:           cnp.APIVersion,
+				logfields.K8sNamespace:            cnp.Namespace,
+				logfields.ServiceID:               svcID,
+			}).Debug("Service updated or deleted, recalculating CiliumNetworkPolicy rules")
+		}
+		initialRecvTime := time.Now()
+
+		resourceID := resourceIDForCiliumNetworkPolicy(key, cnp)
+
+		errs = append(errs, p.resolveCiliumNetworkPolicyRefs(cnp, key, initialRecvTime, resourceID))
+	}
+	return errors.Join(errs...)
+}
+
+// resolveToServices translates all ToServices rules found in the provided CNP
+// and to corresponding ToCIDRSet rules. Mutates the passed in cnp in place.
+func (p *PolicyWatcher) resolveToServices(key resource.Key, cnp *types.SlimCNP) {
+	// We consult the service cache to obtain the service endpoints
+	// which are selected by the ToServices selectors found in the CNP.
+	p.svcCache.ForEachService(func(svcID k8s.ServiceID, svc *k8s.Service, eps *k8s.Endpoints) bool {
+		if !p.isSelectableService(svc) {
+			return true // continue
+		}
+
+		// svcEndpoints caches the selected endpoints in case they are
+		// referenced more than once by this CNP
+		svcEndpoints := newServiceEndpoints(svcID, svc, eps)
```

Comment:
Do we even need the `svcEndpoints` object any more? It looks like `processRule()` could be just a function now.

Reply:
As it says in the comment: It caches endpoints if they are selected by a policy more than once. It's not necessary for correctness.